### ISSUE TITLE
[BUG] Aliasing noise

### ISSUE THREAD

**Issue by sdhizumi** [yokemura/Magical8bitPlug2#18]
**Describe the bug**
Magical8bitPlug2 has an aliasing noise, increase or decrease depending on the frequency rate of the host.
I think this is not a correct sound of an old 8bit game console.

**To Reproduce**
1. Open project (I checked 44.1kHz, 48kHz, and 192kHz project)
2. Load M8P2
3. Play something note

**Expected behavior**
Played something note but it includes an aliasing noise.
This noise affects the impression of the high frequencies, decreases with higher sampling rates.

I tried to compare it with the actual Gameboy's waveform. (Sorry I can't test with FC/NES quickly.)

**Screenshots**
All images captured by Voxengo SPAN Plus with these settings:
Window = HANN
Overlap = 50.0
Avg Time = 500
Type = RT AVG
Block Size = 2048

ANTI-ALIAS = OFF
ALIGN 0 DB = OFF
Offset = OFF
Smooth = OFF

Freq Lo = 20.0
Freq Hi = 20.0k
Range Lo = -96.0
Range Hi = 0.0
Slope = 4.50
(Probably it can be setting the same settings on SPAN, free edition.)

Played 440Hz 50% Pulse note -
on 44.1kHz:
![M8P2_A4_PL50_44 1K_20-20k](https://user-images.githubusercontent.com/31327231/116818017-2c3b4480-aba4-11eb-96ac-0bb1dcb93146.png)
M8P2
![DMG_A4_PL50_44 1K_20-20k](https://user-images.githubusercontent.com/31327231/116818025-35c4ac80-aba4-11eb-991e-5afa389943f0.png)
Gameboy APU

on 48kHz:
![M8P2_A4_PL50_48K_20-20k](https://user-images.githubusercontent.com/31327231/116818035-42490500-aba4-11eb-8a41-82f02ec77549.png)
M8P2
![DMG_A4_PL50_48K_20-20k](https://user-images.githubusercontent.com/31327231/116818038-47a64f80-aba4-11eb-8532-b797e7f6d9a3.png)
Gameboy APU

on 192kHz:
![M8P2_A4_PL50_192K_20-20k](https://user-images.githubusercontent.com/31327231/116818046-555bd500-aba4-11eb-8f07-8c663eb66de6.png)
M8P2
![DMG_A4_PL50_192K_20-20k](https://user-images.githubusercontent.com/31327231/116818051-5a208900-aba4-11eb-8dd4-3ddcc3814eb3.png)
Gameboy APU

NOTE - Gameboy APU has a hissing noise. It looks like this:
![DMG_hiss](https://user-images.githubusercontent.com/31327231/116818424-33fbe880-aba6-11eb-8e14-3655d55ab98c.png)

**Audio**
I uploaded some samples on my drive. It contains a plain text file that wrote some descriptions of samples.
https://drive.google.com/file/d/10BwRPw2CkZUchpEji6YLlRW1o4-M2OeP/view?usp=sharing

**Your environment (please complete the following information):**
 - OS: Windows 10 20H2
 - Host DAW and version: FL Studio 20.8.3, Reaper v6.26
 - Version of Magical 8bit Plug 2: 1.0.0
 - Plugin format of Magical 8bit Plug you used: VST3

(and my friend has tested on Logic, same issue.)

**Additional context**
It looks like anti-aliasing is needed.

**Comment by yokemura** [comment]
Thanks for your info.

Is the waves of M8P2 are rendered directly from DAWs?
Are there any conversion or DA/AD involved?

M8P's audio rendering is no simpler than any waveform generation algorithm.
Just renders -1 or 1 according to the cycle phase.

And as a matter of fact, the waveform rendered directly from DAW(mine is Logic Pro X) looks like this.

<img width="1413" alt="スクリーンショット 2021-05-03 8 16 25" src="https://user-images.githubusercontent.com/384400/116831007-f28c2d00-abe7-11eb-81c3-f2f832ba0cce.png">

I know this is an extreme waveform which is not likely to happen with ordinary instruments, and it's likely that converting it to any other format will cause some distortion as it's too edgy.

**Comment by sdhizumi** [comment]
I updated the samples: added some samples that I exported renderer on FL and Reaper. Please re-download the archive.
Previous samples are recorded by the FL's internal recorder, didn't go through any DA/ADs.

Did you really check my samples?
Probably you (and me) can see almost the same (but a bit different) waveform between DMG_M8P2_44.1K, 48K, and 192K samples in the M8P2 part on Audacity.
It looks like this (these samples are edited to the left channel only):
![image](https://user-images.githubusercontent.com/31327231/117301051-2fd90f00-aeb5-11eb-8c1d-258cb4dae34d.png)

This problem is not in the waveform. I talk about the spectrum.
Please listen and check my samples. You can probably recognize that each sample has the almost same waveform, but will sound different.

**Comment by At-sushi** [comment]
One possible case is that a margin of sampling error causes the alias noises.
I don't know how Gameboy is solving this issue, but I think oversampling may solve the problem.

**Comment by nakakq** [comment]
I'm also interested in this problem, so I'd like to provide some additional information.


> M8P doesn't sample any analog signal but just generate a digital waveform from scratch in pure digital way. So in this context discussing about "aliasing" doesn't make sense.


It seems the M8P's oscillator produces "aliasing" in the sense of *folding noise*.
The following is the spectrogram of the audio `DMG_M8P2_48K.wav` by sdhizumi. The former is GB, the latter is M8P.
The M8P's output has more folding noise.


![image](https://user-images.githubusercontent.com/66908843/131904157-ee4cd97a-d8ad-4a22-a1dd-e536f05b35cd.png)


The implementation in `Source/PulseVoice.cpp` is the following:

```cpp
return angle < rate * MathConstants<float>::pi ? -1.0 : 1.0;
```

The formulation is as same as the definition of an analog square wave (which has infinite harmonic components). Roughly speaking, the implicit harmonics above the `sampleRate / 2` appear as aliases.


Suppression by low-pass/high-pass filtering is difficult because aliasing components already lie between the "true" harmonics. 
To suppress the aliasing of the oscillator, one simple solution is to quantize the waveform's cycle period (and the jumping points) to an integer in sample units. Perhaps this is why there's less aliasing in the GB's square wave. However, this method sacrifices pitch accuracy. Of course, there's a bunch of much more high-quality methods (e.g. [additive synthesis](https://en.wikipedia.org/wiki/Square_wave#Fourier_analysis), [BLIT](https://www.music.mcgill.ca/~gary/307/week5/bandlimited.html), [Poly-BLEP](http://www.martin-finke.de/blog/articles/audio-plugins-018-polyblep-oscillator/), [wavetable with interpolation](https://daichilab.sakura.ne.jp/synthprog/index.html#osc1), ...). 



In my opinion, generating an anti-aliased waveform is reasonable in general. Although, I also like the fat/edgy sound of the current M8P. (I'm wondering if it's possible to add the same aliasing at any sampling rate, even 44.1 kHz or 192 kHz!)

**Comment by yokemura** [comment]
Thanks, @nakakq 

It's an interesting discussion.

It doesn't necessarily means that it deals with sinusoidal waveform only because having the constant PI in the formula.
Simply saying the above code is just checking a timer.

The synthesizer core loops a sort of timing value which ranges from 0 to a certain maximum value. And while the value is under the half of the maximum it produces -1, and then +1 for the rest.  The maximum value is set to 2*PI just because it's *convenient* when you are working with sinusoidal waves, not because there *underlies* a sinusoidal wave.
So in short, it is still generating -1 or 1 according to the timer, not sampling implicit analog waveform.

Saying that, it's also true that the folding happens with above algorithm.  But the note height to make it happen is pretty high.

As @nakakq pointed out, at the point *it* exceeds the `sampleRate / 2` it starts skipping the -1 / +1 sequence.
Then, what is *it* ?  In this case it is note height. 
Supposing the sample rate is 44.1kHz, the note frequency of `sampleRate / 2` is... I can't calculate right away but it's far above C10 (which is the highest value on the table below) 
https://musicinformationretrieval.com/midi_conversion_table.html
Reversely the folding will not happen under this frequency.

To summarize, a PI constant appears in the code but there isn't an implicit sin wave.  The folding still can happen but the note height to make it happen is extremely high.

**Comment by nakakq** [comment]
Maybe my explanation about *implicit harmonics* was unclear, so let me explain in detail.


## Formulation

We're going to generate a square wave (50% duty) with period `P`. The *timer checking* implementation is like this:

<img src="https://render.githubusercontent.com/render/math?math=y%28n%29%3D%5Cbegin%7Bcases%7D-1%2C%26t%3C%7BP%2F2%7D%5C%5C%2B1%2C%26t%5Cge%7BP%2F2%7D%5Cend%7Bcases%7D%7B%5Cquad%7D%5Cmathrm%7Bwhere%7D%5C%2C0%5Cle%7Bn%3CP%7D">

Then, the periodic wave `y(n)` can be expressed as a sum of sine waves (according to [Fourier expansion](https://en.wikipedia.org/wiki/Square_wave#Fourier_analysis)):

<img src="https://render.githubusercontent.com/render/math?math=y%28n%29%3D-%5Cfrac4%5Cpi%5Csum_%7Bh%3D1%7D%5E%5Cinfty%5Cfrac1%7B2h-1%7D%7B%5Csin%282%5Cpi%282h-1%29n%2FP%29%7D">

As you can see, this formula contains **infinite number of sine waves** with frequency multiplied by `2h-1`. These sine waves (except `h = 1`) are so-called overtones (or harmonics). These overtones are the matter. The overtones which exceed the `sampleRate / 2` appear as aliases in a digital signal.  The aliasing occurs even the note frequency is low, because of the infinite number of overtones. To prevent aliasing, we need to limit the number of overtones (so that the all of them don't exceed the `sampleRate / 2`).


## Demonstration

I wrote a C++ program to demonstrate the aliasing effect ([source code](https://gist.github.com/nakakq/cc63cd95ac88473bd96ade8dbc0b2565)). This program outputs 3 WAV files; they are all square waves but are generated by different methods. The note frequency gradually changes from C3 (130.81 Hz) to C10 (16744.04 Hz). The `sampleRate / 2` is 22050 Hz.

The 3 generation methods are:

- `M8P2`: Oscillator from Magical 8-bit Plug 2.
- `Sines`: Sum of sines below the `sampleRate / 2`.
- `ManySines`: Sum of sines, including above the `sampleRate / 2` (below 2.5 MHz).

Results (spectrogram): 

![image](https://user-images.githubusercontent.com/66908843/132297605-39987690-a2aa-4c64-ba67-d23096324745.png)

The spectrogram of `M8P2` and `ManySines` are very similar. This implies the `M8P2` waveform has a lot of overtones which exceeds the `sampleRate / 2`.
In contrast, `Sines` doesn't have aliases; because it limits the overtones to ensure the sampling theorem.

Here's a graphical explanation for the folding of the overtones:

![image](https://user-images.githubusercontent.com/66908843/132297691-679ea5b2-06a9-4b05-acc4-e9c403610348.png)


To sum up, the implicit overtones make the aliasing. I hope this comment is useful!

**Comment by yokemura** [comment]
Got it.
Here's a question.
How is the GBA's waveform seen in the spectrogram?  Does it look like the one of sine wave?

**Comment by yokemura** [comment]
Can I see the waveform of "many sines"?

**Comment by nakakq** [comment]
Here's the original files: [pulse.zip](https://github.com/yokemura/Magical8bitPlug2/files/7146987/pulse.zip)

My spectrogram settings in Audacity (the window size was 1024):

![image](https://user-images.githubusercontent.com/66908843/132931302-613f5daf-5f14-480c-a712-3e299fae4d3f.png)

**Comment by flubber2077** [comment]
@yokemura [here is some info on reducing aliasing waves](https://www.reddit.com/r/synthdiy/comments/cnspor/comment/ewdo1bv/?utm_source=share&utm_medium=web2x&context=3). essentially, some trickery needs to be done at the discontinuities, otherwise harmonics will reflect back down from half the sampling rate, becoming weird inharmonics. id recommend polyblep; its computationally quick and usually the fastest to code into the oscillators

**Comment by yokemura** [comment]
@nakakq Sorry for my long silence. I've been busy in making my music.

Before moving forward the discussion we have to think about what NOISE is, what DISTORTION is.
As far as taking a glance at the waveforms, the one which has noise is "Sines" waveform.
Which waveform should we respect?

<img width="1506" alt="スクリーンショット 2021-12-21 16 11 38" src="https://user-images.githubusercontent.com/384400/146887325-ab300171-c8d5-4628-bc07-4484f9dc9849.png">

And as another point of view, we have to be aware that Sampling Theorem also applies to the analysis side.

My understanding is like this:
- The PURE square waveform is too sharp to be expressed as a convolution of sine waves.
- The reason why m8p can produce such a sharp waveform is that m8p's wave generation is not based on any sampling.
- That's why the spectrogram of m8p waveform has the aliasing. In other words, the wave is too pure to express as a spectrogram.
- The "Sine" waveform is not a square wave but something alike that can avoid the ANALYSIS noise.  Or it can be said as DISTORTED when we put focus on the pure square wave.

**Comment by nakakq** [comment]
@yokemura Thank you! I'd like to reply line-by-line.

In a nutshell:
- Some sort of artifact will be inevitable in digital signal.
- The m8p oscillator: suppress artifacts in **time domain** (clean waveform).
- Alias-free oscillators (e.g. "Sines"): suppress artifacts in **frequency domain** (clean spectrum).
- Human auditory is in frequency domain; Suppressing artifacts in **frequency domain** is reasonable to improve sound quality.

---

> As far as taking a glance at the waveforms, the one which has noise is "Sines" waveform.
> Which waveform should we respect?

In my opinion, it's difficult to discuss the sound quality using the time-domain waveform.

Human hears mainly in the frequency domain (recall the cochlea of the ear), so we should evaluate them on the frequency domain.
The time-domain waveform can be important for evaluating envelope or dynamics, but it doesn't matter for oscillators.

The m8p's waveform is cleaner than the "Sines" **in the time-domain**, but I think the frequency-domain characteristics are more important for the sound.

---

> And as another point of view, we have to be aware that the analysis process can produce noise. In short, analysis may lie. The analysis methods like spectrogram uses FFT/wavelet-based algorithm involve the process to decompose the waveform into sine waves, and it means any waveforms that is difficult to represent as a sinusoidal composition can result in an unexpected expression.

> The PURE square waveform is too sharp to be expressed as a convolution of sine waves.
> That's why m8p's spectrogram shows the impurity. It must be caused by the ANALYSIS process.

I agree that discrete frequency analysis (like FFT/wavelet) can describes the concept of aliasing.

As I mentioned, the frequency domain analysis is suitable for evaluating sound quality.
Actually, you may hear the obvious aliasing in the m8p sample (in the [audio samples previously sent](https://github.com/yokemura/Magical8bitPlug2/issues/18#issuecomment-917314286)), or playing high-pitched notes with vibrato on m8p.

Also, the sampling theorem is a limitation of digital signal, not a limitation of analysis.
So, any digital waveforms can't represent a "true" square wave with infinity resolution (except for the wavelength with an even samples).

---

> The "Sine" waveform is not a square wave but something alike that can avoid the analysis noise. Or it can be said as DISTORTED when we put focus on the pure square wave.

The "Sine" waveform is equal to a low-passed square wave with cutoff frequency `sampleRate / 2`.
(The low-pass filter eliminates the extra harmonics to ensure the sampling theorem.)

So to speak, this is an "antialiased" (bandlimited) version of a square wave.
In the time-domain waveform, there're some ripples around the edge. They are called [Gibbs effect](https://en.wikipedia.org/wiki/Gibbs_phenomenon) and are often seen in low-passed waveforms.
